### ISSUE TITLE
Is it possible to remove the "plug, unplug, rotate" process in a safe way?

### ISSUE THREAD

**Issue by zenware** [maqp/tfc#2]
Basically, I have a personal goal to "productize" this project. I want to squish it all into some slick package that the user doesn't have to worry about setting it up themselves. However, I can't think of a good way aside from maybe something asinine like expanding upon the data-diode concept and making clever use of a DIP switch. However, the possibility of the switches being left in the wrong position, or fiddled with post-infiltration gives me some bad feelings about it. Would be nice if there was some way to do some slicker low-risk deployment.

**Comment by maqp** [comment]
Are you referring to [NH bypass during local key delivery](https://camo.githubusercontent.com/52b0492bdc8fab02f9b479f6582a782681f5c2c9/68747470733a2f2f63732e68656c73696e6b692e66692f752f6f6f7474656c612f74666377696b692f686f775f746f5f7573652f345f6e685f6279706173735f72656e6465722e6a7067)?

This could be done if you can find a multi-position switch that with a single control, sequentially

1. Disconnects all three NH TTL adapter leads from NH-RxM optocoupler

2. Disconnects all three TxM TTL adapter leads from TxM-NH optocoupler

3. Connects TxM TTL adapter to NH-RxM optocoupler's Tx-side

This ensures there is no way for user to perform things in the wrong order. The switch packaging probably needs a custom component so make the casing transparent to allow easy inspection. If you're going this way, please also invest time in finding a suitable (i.e. fast) LED-photodiode pair. By replacing the DIP optocoupler with two-lead components, you don't have to worry about the interdiction problem. Also note that data diode is a patented invention in US ([See last page](http://homepage.divms.uiowa.edu/~jones/voting/diode/RS232tech.pdf)).

Then again, I don't think adding the switch increases security if the alternative is a significantly more complex data diode circuit. NH bypass was never necessary, but as it was an easy way to increase bootstrap security (previous generation of RS232 data diodes could be moved around easily), I went for it.

The current TTL data diode design is more fragile so it's less feasible to do. As it's impossible for me to know if users want to keep using the feature, I haven't removed it as it's only done during the initial local key setup, and because skipping the step is only two presses of `enter` more effort.

**Comment by zenware** [comment]
So far I've drawn up a board to make creating your optocoupler design a little quicker and neater, supposing one thinks that soldering electronics on a board is quicker than pushing them through a breadboard.

Future, plans are to draw up a board that can supply power to three of these https://www.pine64.org/?product=sopine-a64 (or some other similar small computers such as a Pi Zero W and two standard Pi Zero devices) hopefully I can come up with something pocketable, maybe a big huge jacket pocket, but pocketable.

My thoughts are, if it's not easily portable (i.e. you have wired three desktop machines this way) it's compromised by default as soon as you leave it behind. I imagine you think so as well considering your goals are to keep all the wiring out in the open where it can be easily checked and validated.

I also think the NH should be "wifi-mostly" networked, especially when 802.11u becomes more common/widespread. I like the idea of "roaming" networks making a communication origin being that much more difficult to hunt down.

And I think the device should be 100% tossable if you're afraid someone may soon try to retrieve it from you that you'd be fully comfortable (and capable of) smashing it to bits.

To summarize, I think a tfc "package" should be:
* Very Portable
* Wireless First
* Easily Disposable

**Comment by maqp** [comment]
A PCB with possibly preinstalled sockets for DIPs and TTL-adapters would significantly reduce end user effort. Cutting, stripping and bending the wires is surprisingly time consuming. Some TTL adapter models might have pin connectors pre-soldered, or they might have wires (see below).

The fact SoC manufacturers are slowly introducing wireless into their products is making their use much harder in the future. But if the community is agile and expands TFC to new platforms, this shouldn't be a problem with newer generations of hardware.

Pocketable implementation would be truly interesting to see. The current smallest package I can think is three GPD Pockets and small black box containing the data diode circuit where TTL adapter is [integrated in the USB-plug](https://sc01.alicdn.com/kf/HTB124srGXXXXXXEXpXXq6xXFXXXZ/201978820/HTB124srGXXXXXXEXpXXq6xXFXXXZ.jpg). But since RPi Zero has internal UART, it's even smaller and cheaper a form factor. Please have a look at the [original data diode design](https://imgur.com/a/5Cv19) regardling necessary voltage divider implementation

It's pretty much impossible to carry the hardware with you for the rest of your life. If the anonymity Tor routing provides fails and you end up under targeted, close proximity surveillance, there are too many attacks to defend against. Of course, being on the move while using TFC helps.

>I imagine you think so...

I agree. Another reason is to be able to verify what was delivered in the mail in the first place.

Tossability is important and having data on two microSD cards makes this easy as long as user physically destroys the storage media. For HDD, the new `/wipe` command gives the fastest first line of defense before user can physically shred their drives.

**Comment by maqp** [comment]
The NH (aka Networked Computer) bypass messages is no longer enabled by default. I'm closing this issue as the feature will most likely be deprecated in the near future: Local key delivery and key decryption key entry should only be performed in places where the chances of being watched is low, or under a [magic mantle of power](https://i.stack.imgur.com/bDeLe.png). 

The protocol will remain compatible so that if the user so chooses, they can re-arrange the TTL-adapters across an optocoupler and deliver the local key from TCB half to another, bypassing the Relay Program. Once this done, they can issue the `/localkey` command and once local key has been installed, they can return the TTL-adapters to their original places.
